### ISSUE TITLE
Vitest+Typescript: TypeError: Cannot set property X of #<Y> which has only a getter

### ISSUE THREAD

**Issue by ffknob** [vitest-dev/vitest#1494]
### Describe the bug

I'm having some trouble and need some advice on creating a TS class with private/protected readonly members and have it pass my vitest tests.

asset.ts
```typescript
export enum AssetType {
  STOCK,
  CRYPTO,
}

export abstract class Asset {
  protected readonly _type: AssetType;
  protected readonly _symbol: string;
  protected readonly _name: string;
  protected _amount: number;
  protected readonly _initialPrice: number;

  constructor(
    type: AssetType,
    symbol: string,
    name: string,
    amount: number,
    initialPrice: number,
  ) {
    this._type = type;
    this._symbol = symbol;
    this._name = name;
    this._amount = amount;
    this._initialPrice = initialPrice;
  }

  get type(): AssetType {
    return this._type;
  }

  get symbol(): string {
    return this._symbol;
  }

  get name(): string {
    return this._name;
  }

  get amount(): number {
    return this._amount;
  }

  set amount(amount: number) {
    this._amount = amount;
  }

  get initialPrice(): number {
    return this._initialPrice;
  }
}

```

stock.ts:
```typescript
import { Asset, AssetType } from '@core/asset';

export class Stock extends Asset {
  constructor(
    symbol: string,
    name: string,
    amount: number,
    initialPrice: number,
  ) {
    super(AssetType.STOCK, symbol, name, amount, initialPrice);
  }
}

```

broker.test.ts
```typescript
import { describe, it, expect, beforeAll } from 'vitest';
import { Broker } from '@core/broker';
import { Position } from '@core/position';
import { Stock } from '@core/stock';

let stock: Stock;

beforeAll(() => {
  stock = new Stock('XYZ', 'XYZ Co.', 1000, 1.0);
});

describe('addPosition', () => {
  it("should add position to the broker's positions", () => {
    const positions: Position[] = [];
    const broker: Broker = new Broker('John Smith', positions);
    const newPosition: Position = new Position(stock, 1);

    expect(broker.positions).toEqual([newPosition]);
  });
});

```

Result after running `vitest`:

```console
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Suites 1 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 FAIL  src/core/broker/broker.test.ts > addPosition
TypeError: Cannot set property asset of #<Position> which has only a getter
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/2]⎯

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Tests 1 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 FAIL  src/core/broker/broker.test.ts > addPosition > should add position to the broker's positions

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Error ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
TypeError: Cannot read properties of undefined (reading 'stacks')
 ❯ parseStacktrace file:/home/ffknob/Entwicklung/workspace/ffknob/x-udemy-js-unit-testing/node_modules/vitest/dist/chunk-utils-timers.8a5e7cd5.js:5678:9
 ❯ printError file:/home/ffknob/Entwicklung/workspace/ffknob/x-udemy-js-unit-testing/node_modules/vitest/dist/chunk-vite-node-externalize.92c54acc.js:10045:18
 ❯ Vitest.printError file:/home/ffknob/Entwicklung/workspace/ffknob/x-udemy-js-unit-testing/node_modules/vitest/dist/chunk-vite-node-externalize.92c54acc.js:10623:12
 ❯ VerboseReporter.printTaskErrors file:/home/ffknob/Entwicklung/workspace/ffknob/x-udemy-js-unit-testing/node_modules/vitest/dist/chunk-vite-node-externalize.92c54acc.js:8640:22
 ❯ VerboseReporter.reportSummary file:/home/ffknob/Entwicklung/workspace/ffknob/x-udemy-js-unit-testing/node_modules/vitest/dist/chunk-vite-node-externalize.92c54acc.js:8591:18
 ❯ processTicksAndRejections node:internal/process/task_queues:96:5
 ❯ async VerboseReporter.onFinished file:/home/ffknob/Entwicklung/workspace/ffknob/x-udemy-js-unit-testing/node_modules/vitest/dist/chunk-vite-node-externalize.92c54acc.js:8486:5
 ❯ async VerboseReporter.onFinished file:/home/ffknob/Entwicklung/workspace/ffknob/x-udemy-js-unit-testing/node_modules/vitest/dist/chunk-vite-node-externalize.92c54acc.js:9266:5
 ❯ async Vitest.report file:/home/ffknob/Entwicklung/workspace/ffknob/x-udemy-js-unit-testing/node_modules/vitest/dist/chunk-vite-node-externalize.92c54acc.js:10574:5


```

So, since I'm not the one setting the property explicit with a setter, I guess that's something vitest is doing somehow...

I can manage it to work if I remove the `readonly` modifier and create a setter for the propertuy, but I wonder if I can pass this error without having to make changes to my code.

Any thoughts?

Thank you

### Reproduction

(see example code above)

### System Info

```shell
Need to install the following packages:
  envinfo
Ok to proceed? (y) 

  System:
    OS: Linux 5.17 openSUSE Tumbleweed 20220506
    CPU: (8) x64 Intel(R) Core(TM) i7-10510U CPU @ 1.80GHz
    Memory: 3.84 GB / 15.41 GB
    Container: Yes
    Shell: 5.8 - /usr/bin/zsh
  Binaries:
    Node: 16.14.0 - ~/.nvm/versions/node/v16.14.0/bin/node
    Yarn: 1.22.18 - /usr/bin/yarn
    npm: 8.3.1 - ~/.nvm/versions/node/v16.14.0/bin/npm
  Browsers:
    Chrome: 101.0.4951.54
    Firefox: 100.0
  npmPackages:
    vitest: ^0.10.4 => 0.10.4 
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by sheremet-va** [comment]
I think this might be a problem with how we handle Errors. We need to serialize them, so we clone it's properties, and that's why it might happen. It happens here, if someone is interested to fix this:

https://github.com/vitest-dev/vitest/blob/59aa8e523148c5a15c9aa2eae86d21c3705f894a/packages/vitest/src/runtime/error.ts#L8

**Comment by Tchoupinax** [comment]
Hello,
I got the same issue for several tests, one exemple

```
Error: Failed to fully serialize error: Method get Set.prototype.size called on incompatible receiver #<Set>
Inner error message: Validation error
```

vitest : 0.15.1
node : v16.15.1

**Comment by benbender** [comment]
I'm seeing the same error when using `vitest`, https://github.com/capricorn86/happy-dom/ & https://github.com/testing-library/jest-dom.

I thought this is an issue in `happy-dom`, but it may be better resolved here.

Related: https://github.com/capricorn86/happy-dom/issues/544 & https://github.com/capricorn86/happy-dom/pull/545

**Comment by ChrisTowles** [comment]
I tried to recreate this issue without any luck, not all the required files are listed so its not clear.

![image](https://user-images.githubusercontent.com/3102683/185823945-60af1b87-572c-4ac9-84a0-17db03ca52bc.png)


[Edit on StackBlitz ⚡️](https://stackblitz.com/edit/vitest-dev-vitest-fckmry?file=README.md,test%2Fbroker.test.ts)

And repo at
https://github.com/ChrisTowles/vitest-dev-vitest-fckmry

**Comment by hnrq** [comment]
I've been getting this error everywhere when using `happy-dom` with vitest. `jsdom` works smoothly on the other hand

**Comment by sheremet-va** [comment]
Fixed with #1921

**Comment by hnrq** [comment]
@sheremet-va any guess on when the next release will be out? :pleading_face:
